lib/class-wp-theme-json-schema-gutenberg.php

```diff
@@ -0,0 +1,196 @@
+<?php
+/**
+ * WP_Theme_JSON_Schema_Gutenberg class
+ *
+ * @package WordPress
+ * @subpackage Theme
+ * @since 5.9.0
+ */
+
+if ( class_exists( 'WP_Theme_JSON_Schema_Gutenberg' ) ) {
+	return;
+}
+
+/**
+ * Class that migrates a given theme.json structure to the latest schema.
+ *
+ * This class is for internal core usage and is not supposed to be used by extenders (plugins and/or themes).
+ * This is a low-level API that may need to do breaking changes. Please,
+ * use get_global_settings, get_global_styles, and get_global_stylesheet instead.
+ *
+ * @since 5.9.0
+ * @access private
+ */
+#[AllowDynamicProperties]
+class WP_Theme_JSON_Schema_Gutenberg {
+
+	/**
+	 * Maps old properties to their new location within the schema's settings.
+	 * This will be applied at both the defaults and individual block levels.
+	 */
+	const V1_TO_V2_RENAMED_PATHS = array(
+		'border.customRadius'         => 'border.radius',
+		'spacing.customMargin'        => 'spacing.margin',
+		'spacing.customPadding'       => 'spacing.padding',
+		'typography.customLineHeight' => 'typography.lineHeight',
+	);
+
+	/**
+	 * Function that migrates a given theme.json structure to the last version.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param array $theme_json The structure to migrate.
+	 *
+	 * @return array The structure in the last version.
+	 */
+	public static function migrate( $theme_json ) {
+		if ( ! isset( $theme_json['version'] ) ) {
+			$theme_json = array(
+				'version' => WP_Theme_JSON::LATEST_SCHEMA,
+			);
+		}
+
+		if ( 1 === $theme_json['version'] ) {
+			$theme_json = self::migrate_v1_to_v2( $theme_json );
+		}
+
+		if ( 2 === $theme_json['version'] ) {
+			$theme_json = self::migrate_v2_to_v3( $theme_json );
+		}
+
+		return $theme_json;
+	}
+
+	/**
+	 * Removes the custom prefixes for a few properties
+	 * that were part of v1:
+	 *
+	 * 'border.customRadius'         => 'border.radius',
+	 * 'spacing.customMargin'        => 'spacing.margin',
+	 * 'spacing.customPadding'       => 'spacing.padding',
+	 * 'typography.customLineHeight' => 'typography.lineHeight',
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param array $old Data to migrate.
+	 *
+	 * @return array Data without the custom prefixes.
+	 */
+	private static function migrate_v1_to_v2( $old ) {
+		// Copy everything.
+		$new = $old;
+
+		// Overwrite the things that changed.
+		if ( isset( $old['settings'] ) ) {
+			$new['settings'] = self::rename_paths( $old['settings'], self::V1_TO_V2_RENAMED_PATHS );
+		}
+
+		// Set the new version.
+		$new['version'] = 2;
+
+		return $new;
+	}
+
+
+	/**
+	 * Migrate away from the previous syntax that used a top-level "behaviors" key
+	 * in the `theme.json` to a new "lightbox" setting.
+	 *
+	 * @param array $old Data with (potentially) behaviors.
+	 * @return array Data with behaviors removed.
+	 */
+	private static function migrate_v2_to_v3( $old ) {
```

Comment:
I'm uncertain about one thing regarding the migration.

The migration that @michalczaplinski added here _does not_ update the `theme.json` on disk. Does this mean we just leave the [`migrate_v2_to_v3`](https://github.com/WordPress/gutenberg/pull/54071/commits/443aa5b4ed70eda9ae9279bc9e8653543212150d) in place for the foreseeable future?

If that is the case, do we also leave the backwards compatibility for the block attributes in place as well?

I don't see a reason we _need_ to remove them.

Reply:
> The migration that @michalczaplinski added here _does not_ update the `theme.json` on disk.

That is correct. 

> Does this mean we just leave the [migrate_v2_to_v3](https://github.com/WordPress/gutenberg/pull/54071/commits/443aa5b4ed70eda9ae9279bc9e8653543212150d) in place for the foreseeable future?

Maybe. Since this change never made it to core, we could potentially add another `_deprecated_function()` call to the migration to instruct users to update their theme.json like we have for the block supports that require updating their block.json.

However, users were able to override the theme.json settings with the UI, right? If so, we'd need a way to migrate those changes in the database to the new format too before this point.

If we can do all that, then maybe we don't actually need a full v3. Instead we could have this v2 with lightbox behaviors to v2 with lightbox settings based on the presence of `behaviors` that never leaves Gutenberg and gets removed in two versions like the rest.

Hopefully that make sense—I'm kinda just thinking out loud here. Maybe @jorgefilipecosta can chime in.

Reply:
> However, users were able to override the theme.json settings with the UI, right?

This is correct.

> If so, we'd need a way to migrate those changes in the database to the new format too before this point.

Ok, got it. I imagine there must be some mechanism in Gutenberg for doing this 🤔

Reply:
> However, users were able to override the theme.json settings with the UI, right? If so, we'd need a way to migrate those changes in the database to the new format too before this point.

As far as I can tell, those changes should migrate just fine with just the `migrate_v2_to_v3` function! I've just tested it and it seems fine. 

Also, e.g. the `gutenberg_get_global_settings()` function uses `WP_Theme_JSON_Gutenberg` internally so it uses the migrated settings.

Reply:
> As far as I can tell, those changes should migrate just fine with just the `migrate_v2_to_v3` function! I've just tested it and it seems fine.

Right, but to @artemiomorales's point this has to be a permanent addition to the schema class that gets ported to core even if behaviors never made it. If there was a way to update the database and run the migration without updating the version, we could skip porting this code to core. It feels nicer to me if we don't need to have references to behaviors in core if possible, but this certainly isn't a blocker in my mind.